### PR TITLE
hoc2023: AI field

### PR DESCRIPTION
**Pull request by breville**
This change renders the three chosen input emoji in the custom field of the new AI block.

It also removes the generation of dancers, and shortens the block text.

Screenshot 2023-10-08 at 10 43 07 PM

### CODE REVIEW

**Comment by breville** [comment]
Currently, the block field value is a JSON string like this: 
```
'{"inputs": ["sparkle", "robot", "warm"], "backgroundEffect": "sparkles", "backgroundColor": "warm", "foregroundEffect": "color_lights"}'
```
By including the inputs, the custom field implementation can represent the emoji chosen in the generating phase.  The other values capture what the AI generated from these inputs.

@code-dot-org/code-tools, might it be better to use a [mutator](https://developers.google.com/blockly/guides/create-custom-blocks/extensions#mutators) to store some of the values?

**Review by molly-moen — COMMENTED** [review]
If the state that needs to be serialized is only relevant to the custom field, I think you don't need a mutator ([source](https://developers.google.com/blockly/guides/configure/web/serialization#when_to_save_extra_state)). Does that seem like that covers this situation?

File: apps/src/dance/ai/DanceAiModal.tsx [hunk 1]
```diff
@@ -112,24 +112,27 @@ const DanceAiModal: React.FunctionComponent<DanceAiProps> = ({onClose}) => {
...
+    startAi(inputs, request);
     setMode(Mode.GENERATING);
   };
 
-  const startAi = async (value: string) => {
+  const startAi = async (inputs: string[], value: string) => {
     const responseJsonString = await doAi(value);
     const response = JSON.parse(responseJsonString);
 
     // "Pick" a subset of fields to be used.  Specifically, we exclude the
```
Comment (bencodeorg) [inline apps/src/dance/ai/DanceAiModal.tsx:128]: Is there a way to do this that just grabs the properties you want from the response, then combines with the inputs separately? A little tricky to read as-is. I could also imagine this object containing two keys, `inputs` and `outputs`, with `outputs` having the backgroundEffect, etc. keys, but that's probably out of scope for this PR. At least, I think the comment above `pickedResponse` should be updated.

File: apps/src/dance/ai/cdoFieldDanceAi.js [hunk 2]
```diff
@@ -123,35 +96,41 @@ class CdoFieldDanceAi extends GoogleBlockly.Field {
...
-        height: FIELD_HEIGHT,
+        width: 3 * ITEM_WIDTH + 2 * FIELD_PADDING,
+        height: FIELD_HEIGHT + 2,
         rx: 3,
       },
       this.backgroundElement
     );
 
+    inputs?.forEach((input, index) => {
+      GoogleBlockly.utils.dom.createSvgElement(
```
Comment (bencodeorg) [inline apps/src/dance/ai/cdoFieldDanceAi.js:107]: This is old, but we can't just use the global `Blockly` import here?

File: apps/src/dance/ai/cdoFieldDanceAi.js [hunk 3]
```diff
@@ -123,35 +96,41 @@ class CdoFieldDanceAi extends GoogleBlockly.Field {
...
+          href: `/blockly/media/dance/ai/emoji/${input}.svg`,
+        },
+        this.backgroundElement
+      );
+    });
+
     // Now attach the text element to the background parent.  It will
     // render on top of the background rectangle.
-    this.backgroundElement.appendChild(textElement);
+    //this.backgroundElement.appendChild(textElement);
```
Comment (bencodeorg) [inline apps/src/dance/ai/cdoFieldDanceAi.js:121]: Remove?

**Review by bencodeorg — APPROVED** [review]
A couple nits but looks great!

**Comment by breville** [comment]
> If the state that needs to be serialized is only relevant to the custom field, I think you don't need a mutator ([source](https://developers.google.com/blockly/guides/configure/web/serialization#when_to_save_extra_state)). Does that seem like that covers this situation?

Good point.  I'm still a little hazy, but it does seem reasonable to say that this custom field's state is genuinely a combination of the input emoji and AI-generated output, and we can tweak our field rendering to only show the former and the code generation to only show the latter, as we want.  It's a bit unusual, but that might be a product of this particular block's unique design.

File: apps/src/dance/ai/cdoFieldDanceAi.js (hunk 2 above)
Comment (breville) [reply apps/src/dance/ai/cdoFieldDanceAi.js:107]: Good question.  We can, and that will use the `BlocklyWrapper`.  We also refer to `GoogleBlockly.Field` in this file, so I'm curious if it's appropriate to reference both `Blockly` and `GoogleBlockly` in the same file, for different purposes?  (Maybe @code-dot-org/code-tools have thoughts too.)

Comment (molly-moen) [reply apps/src/dance/ai/cdoFieldDanceAi.js:107]: As a general rule I think we should use the wrapper. The reason I can think of to not use it is if we want to use the google blockly version of a parameter/function/etc. and we have our own custom version. But I think that should rarely happen outside of the wrapper class itself.

File: apps/src/dance/ai/DanceAiModal.tsx (hunk 1 above)
Comment (breville) [reply apps/src/dance/ai/DanceAiModal.tsx:128]: > Is there a way to do this that just grabs the properties you want from the response, then combines with the inputs separately? 

Agreed, and hopefully improved.

> I could also imagine this object containing two keys, inputs and outputs, with outputs having the backgroundEffect, etc. keys, but that's probably out of scope for this PR.

Having `inputs` and `outputs` as peers is a great idea.  If this general format sticks, definitely open to doing that in a follow-up.

File: apps/src/dance/ai/DanceAiModal.tsx [hunk 4]
```diff
@@ -117,50 +117,51 @@ const DanceAiModal: React.FunctionComponent<DanceAiProps> = ({onClose}) => {
...
-      response
-    );
-    const pickedResponseJson = JSON.stringify(pickedResponse);
+    }) => ({
+      backgroundEffect,
+      backgroundColor,
+      foregroundEffect,
+    }))(result);
+
+    const fullResult = {inputs, ...result};
```
Comment (fisher-alice) [inline apps/src/dance/ai/DanceAiModal.tsx]: Do you intend `fullResult = {inputs, ...pickedResult};` ?

Comment (breville) [reply apps/src/dance/ai/DanceAiModal.tsx]: Yes, thanks!

**Pull request merged**